### DISCUSSION TITLE
PHP 8.2 and deprecated static::method() calls for PHP 8.2

### DISCUSSION THREAD

**Discussion by Jimmi08** [e107inc/e107#4987] [Q&A Core] [answered]
Just note because I really don't have idea what I am doing and I needed to rid warnings in my PHP 8.2 local installation 

Inspiration in drupal:

- https://wiki.php.net/rfc/deprecate_partially_supported_callables
- [https://wiki.php.net/rfc/partially-supported-callables-expand-deprecatio...](https://wiki.php.net/rfc/partially-supported-callables-expand-deprecation-notices)

```
"self::method"       -> self::class . "::method"
"parent::method"     -> parent::class . "::method"
"static::method"     -> static::class . "::method"
["self", "method"]   -> [self::class, "method"]
["parent", "method"] -> [parent::class, "method"]
["static", "method"] -> [static::class, "method"]
```

Used in prepare_request() in e107_class.php
Warnings are away

This worked:

```
		// Block common bad agents / queries / php issues.
		array_walk($_SERVER,  [self::class, "filter_request"] , '_SERVER');
		if (isset($_GET))
		{
			array_walk($_GET, [self::class, "filter_request"], '_GET');
		}
		if (isset($_POST))
		{
			array_walk($_POST,   [self::class, "filter_request"], '_POST');
			reset($_POST);		// Change of behaviour in PHP 5.3.17?
		}
		if (isset($_COOKIE))
		{
			array_walk($_COOKIE, [self::class, "filter_request"], '_COOKIE');
		}
		if (isset($_REQUEST))
		{
			array_walk($_REQUEST, [self::class, "filter_request"], '_REQUEST');
		}
```

**Comment by Deltik** [comment, marked as answer]
Please make a pull request with the fix!

**Reply by Jimmi08** [reply]
Doesn't it break stuff with older PHP version?

**Reply by Deltik** [reply]
Should be fine as long as we're not using the [first-class callable syntax](https://www.php.net/manual/en/functions.first_class_callable_syntax.php) from PHP 8.1.  We're still targeting a minimum PHP version of 5.6.

**Reply by Jimmi08** [reply]
There is shortened version of array syntax too. Is it not problem?

**Reply by Deltik** [reply]
Yes, it's fine.  The short array syntax was introduced in PHP 5.4.

**Reply by Deltik** [reply]
https://github.com/e107inc/e107/pull/4988, which fixes the deprecation warnings, has been merged!

**Comment by RanaHalajianO** [comment]
It looks like you are referring to the deprecation of partially supported callables in PHP 8.1 and their removal in PHP 8.2. The changes you mentioned involve replacing certain types of callables with their fully qualified equivalents to avoid warnings and errors in the latest PHP version.

In the code you provided, it seems that you used the self, parent, and static keywords in some of your callable functions. To update these, you can replace them with their fully qualified versions:

self::method -> self::class . "::method"
parent::method -> parent::class . "::method"
static::method -> static::class . "::method"
["self", "method"] -> [self::class, "method"]
["parent", "method"] -> [parent::class, "method"]
["static", "method"] -> [static::class, "method"]
The array_walk function is then used to apply the filter_request function to each element of the $_SERVER, $_GET, $_POST, $_COOKIE, and $_REQUEST arrays. The filter_request function seems to be a custom function that filters the input data to prevent common bad agents/queries/php issues.

By updating your callable functions and applying the filter_request function to your input data, you should be able to resolve the warnings in your PHP 8.2 installation.

**Reply by Deltik** [reply]
This looks like a new kind of spam bot that provides plausible-sounding but false information.

I'm leaving this post up because it's fascinating how tailored the response is, but readers should disregard it because it doesn't tell us any new information and has incorrect syntax and wrongly identifies PHP 8.2 rather than PHP 9 as the removal target for partially supported callables.
